### PR TITLE
Bug 2087159: Modify PPC to generate new "offlined" parameter in Performance Profile

### PR DESCRIPTION
**Pull request by jlojosnegros**
A new parameter is going to be added to Performance Profile  by #345. 
This PR aims to modify the Performance Profile Creator to be able to generate it.

To calculate offlined cpus we take into account the following input parameters:
- `offlined-cpu-count` : the number of cpus the user wants to set offline.
- `disable-ht`: Marks if the user want to disable logical processor siblings to disable hyperthreading
- `power-consumption-mode`: values: “default”, “low-latency”, “ultra-low-latency”

if `disable-ht` is true sibling logical processors will not be considered in the calculation of any of the CPUSets.
Unless `power-consumption-mode` is equal to `ultra-low-latency` ( which is considered a high-consumption scenario), we will first try to look for a complete socket to set offline, that is a socket where all the logical processors are still not in any set.
If we still have not set offline enough logical processors to fulfill the `offlined-cpu-count` we will try to set offline as many logical processor siblings as possible
If we still have not set offline enough logical processors to fulfill the `offlined-cpu-count` we will try to set offline any logical processor that is not already in a set.

Depends on #345

### CODE REVIEW

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 1]
```diff
@@ -289,30 +289,108 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+	}
+
+	if offlinedCPUCount >= int(cpuInfo.TotalThreads) {
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the offlined CPU count in the range [0,%d]", cpuInfo.TotalThreads-1)
 	}
+
+	if reservedCPUCount+offlinedCPUCount >= int(cpuInfo.TotalThreads) {
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved plus offlined CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:310]: shouldn't this range be `[0, len(cpuInfo.TotalThreads)-reservedCPUCount-1]` ?

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 2]
```diff
@@ -289,30 +289,108 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
 	}
+
 	htEnabled, err := ghwHandler.IsHyperthreadingEnabled()
 	if err != nil {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't determine if Hyperthreading is enabled or not: %v", err)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't determine if Hyperthreading is enabled or not: %v", err)
+	}
+
+	//Calculate offlined cpus
+	offlined, err := ghwHandler.GetOfflinedCPUs(cpuInfo, topologyInfo, offlinedCPUCount, disableHTFlag, htEnabled)
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:325]: so offlined takes precedence over reserved and isolated when computing cpu set partitioning, right?
Would it be simpler (and safer) if we compute first the reserved set, then the offlined set and then we trivially (it seems trivial at least) compute the isolated set as the remaining CPU set?

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 3]
```diff
@@ -289,30 +289,108 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
+
+	if lpOfflined < offlinedCPUCount {
+		log.Warnf("could not offline enough logical processors (required:%d, offlined:%d)", offlinedCPUCount, lpOfflined)
 	}
+	return offlinedCPUs.Result(), nil
+}
+
+func (ghwHandler GHWHandler) GetReservedCPUs(offlinedCPUs cpuset.CPUSet, topologyInfo *topology.Info, reservedCPUCount int, splitReservedCPUsAcrossNUMA bool, disableHTFlag bool, htEnabled bool) (cpuset.CPUSet, *topology.Info, error) {
+
+	// TODO maybe it is not a good idea to modify topoly information inside this function
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:386]: Altering the topology data provided by the GHW is not wrong per se, but seems a smell. However it's possible that avoiding so would require significant refactoring (I can't say, but can't rule out either)

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 1 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:310]: yeah you are right

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 2 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:325]: I have nothing against doing it as you propose but I do not understand why it would be simpler and safer to calculate first reserved. 

I thought of calculating offlined first because:
- if user do not want to disable HT and offlinedCount is enough to disable a complete socket we can do it to save some power ( as far as I understood @marioferh  measurements that's the only scenario where some power saving occurs)
- otherwise I will try to offline the sibling threads to help with performance.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 3 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:398]: Yeah, that's true. I think we could revisit this after this feature is done.

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 4]
```diff
@@ -346,15 +438,23 @@ func newCPUAccumulator() *cpuAccumulator {
...
+func (ca *cpuAccumulator) AddCoresWithFilter(max int, cores []*cpu.ProcessorCore, filterLogicalProcessor func(int, int) bool) int {
+	initialCount := ca.count
 	for _, processorCore := range cores {
-		for _, core := range processorCore.LogicalProcessors {
+		for index, logicalProcessorId := range processorCore.LogicalProcessors {
 			if ca.count < max || max == allCores {
-				ca.builder.Add(core)
-				ca.count++
+				if filterLogicalProcessor(index, logicalProcessorId) {
+					ca.builder.Add(logicalProcessorId)
```
Comment (jlojosnegros) [inline pkg/performanceprofile/profilecreator/profilecreator.go:451]: I think there could be an error here with cpuAccumulator.
If `logicalProcessorId` is already present in the builder it will be updated but the number of elements will not be increased therefore `ca.count` will have a wrong value.

If `build.Add` could return the number of elements added this would help ... otherwise I think ca.count should be calculated as `builder.Size` to avoid the problem.
@fromanirh, @Tal-or @swatisehgal  what do you think?

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 1 above)
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:325]: well, the offline amount is anyway bound between 1 and (total - reserved), so it seems nicer to let this limitation be visible in the error message that, unless I'm missing something, can otherwise be misleading.
And yes, we totally need to have at least 1 reserved core

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 2 above)
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:340]: the reasoning is that reserved is compulsory, and the reserved set is the most important set because it's needed for the node to be functional. IOW, without proper setting reserved set, we can kill the node, which is basically among the worst possible outcomes.

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 5]
```diff
@@ -289,30 +304,105 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
+}
+
+func (ghwHandler GHWHandler) GetOfflinedCPUs(cpuInfo *cpu.Info, topologyInfo *topology.Info, offlinedCPUCount int, disableHTFlag bool, htEnabled bool) (cpuset.CPUSet, error) {
+	offlinedCPUs := newCPUAccumulator()
+	lpOfflined := 0
+
+	//currently HT is enabled on the system and the user wants to disable HT
+	if htEnabled && disableHTFlag {
+		for _, node := range topologyInfo.Nodes {
+			lpOfflined += offlinedCPUs.AddCoresWithFilter(offlinedCPUCount, node.Cores, filterFirstLogicalProcessorInCore)
```
[the remark this answers is not in the document]
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:374]: Well I do not know If this is an error perse ... I mean I have chosen not to keep offlining cpus as if `disable-ht` parameter has "precedence" over the number of cpus to disable, but I do not know if that assumption is correct

[the remark this answers is not in the document]
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:374]: uhm, let's just clarify somewhere (commit message, code comments, design doc, you name it) how we want the options to interact and let's be consistent.
I don't actually remember how PPC treats HT disabling.

**Comment by jlojosnegros** [comment]
just rebased over master

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:374]: Ok I have tried to clarify both the parameters and the decision flow in the design document. 
https://docs.google.com/document/d/1ZYs2CD9dfdXECzubpnpdkZXQtjajvhdKJ_jb28h_V4c/edit?pli=1#heading=h.5h27mxaz7he0

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 4 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:451]: As this is an offline tool and I do not think we expect cpuAccumulator to have thousands of elements, could we go with using ca.builder.Size() to update ca.count in the meantime?

Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:451]: I guess we can. I was (and still am) thinking more about code clarity (`cpuset.Builder` reuse is effectively doing more harm than good now) rather than performance.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 2 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:340]: ok, let me then change the order in which they are calculated.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 4 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go]: Make a change proposal here : https://github.com/openshift/cluster-node-tuning-operator/pull/354/commits/954adafefd73bcae947dcaaeea1c745d6a82687e

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 6]
```diff
@@ -334,31 +334,53 @@ func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, sp
...
+		done:  false,
 	}
 }
 
 // AddCores adds logical cores from the slice of *cpu.ProcessorCore to a CPUset till the cpuset size is equal to the max value specified
 // In case the max is specified as allCores, all the cores from the slice of *cpu.ProcessorCore are added to the CPUSet
-func (ca *cpuAccumulator) AddCores(max int, cores []*cpu.ProcessorCore) {
+func (ca *cpuAccumulator) AddCores(max int, cores []*cpu.ProcessorCore) int {
+	if ca.done {
+		return -1
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:491]: let's return the error condition out of band:
```
func (ca *cpuAccumulator) AddCores(max int, cores []*cpu.ProcessorCore) (int, error) {
	if ca.done {
		return -1, fmt.Errorf("CPU accumulator finalized")
        }
        ...
}
```

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 7]
```diff
@@ -334,31 +334,53 @@ func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, sp
...
-				ca.count++
+				_, found := ca.elems[logicalProcessorId]
+				ca.elems[logicalProcessorId] = struct{}{}
+				if !found {
+					ca.count++
+				}
 			}
 		}
 	}
+	return ca.count - initialCount
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:507]: this is covered by existing tests, right?

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 8]
```diff
@@ -334,31 +334,53 @@ func (ghwHandler GHWHandler) GetReservedAndIsolatedCPUs(reservedCPUCount int, sp
...
+}
+
+func (ca *cpuAccumulator) mapKeys() []int {
+	keys := make([]int, len(ca.elems))
+	i := 0
+	for k := range ca.elems {
+		keys[i] = k
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:514]: maybe just use append? also, we can probably inline this function in `Result`
```
func (ca *cpuAccumulator) Result() cpuset.CPUSet {
        var keys []int
        for k := range ca.elems {
             keys = append(keys, k)
        }
	return cpuset.NewCPUSet(keys...)
}
```

File: cmd/performance-profile-creator/cmd/root.go [hunk 9]
```diff
@@ -465,6 +472,7 @@ func getProfileData(args ProfileCreatorArgs, cluster ClusterData) (*ProfileData,
 	kernelArgs := profilecreator.GetAdditionalKernelArgs(args.DisableHT)
 	profileData := &ProfileData{
 		reservedCPUs:           reservedCPUs.String(),
+		offlinedCPUs:           "", //todo assing value when calculated
```
Comment (fromanirh) [inline cmd/performance-profile-creator/cmd/root.go]: OK, this (and the comment above) are fixed in the next commit. In the future please consider fixing them in the same commit

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 10]
```diff
@@ -289,30 +304,121 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
 	cpuInfo, err := ghwHandler.CPU()
 	if err != nil {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain CPU info from GHW snapshot: %v", err)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("can't obtain CPU info from GHW snapshot: %v", err)
 	}
 
+	// Check limits are in range
 	if reservedCPUCount <= 0 || reservedCPUCount >= int(cpuInfo.TotalThreads) {
-		return cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
+		return cpuset.CPUSet{}, cpuset.CPUSet{}, cpuset.CPUSet{}, fmt.Errorf("please specify the reserved CPU count in the range [1,%d]", cpuInfo.TotalThreads-1)
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:342]: this should be addressed in a later PR, however is too early in the flow to target the user in the error message ("please specify..."). At this layer we should just report the error, and let another layer treat with the user interface

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 3 above)
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:439]: typo: `topoly`

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 11]
```diff
@@ -242,6 +242,31 @@ func (ghwHandler GHWHandler) CPU() (*cpu.Info, error) {
 	return ghw.CPU(ghwHandler.snapShotOptions)
 }
 
+func (ghwHandler GHWHandler) SortedCPU() (*cpu.Info, error) {
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:245]: this feature should be proposed to GHW - we should file an issue at least.
Additionally, I cant' help but keep wondering if instead of mangling `cpu.Info` we shoudl just compute here the representation we need, which seems a simple `[]int`.
IOW, instead of

1. computing a new `cpuInfo` with the sorted values
2. iterate over the mangled `cpuInfo` to extract the cpuid slice
3. use the cpuid slice

just compute the cpuid slice here and use it directly

**Review by fromanirh — COMMENTED** [review]
partial review, need to get to the allocation logic proper

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 11 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:245]: Sorry but I thin I do not fully understand your comment. 

we need all the socket -> core -> logical processor structure if we want to use that information to selectively disable some processors.

Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:245]: right, we need to keep some hierarchy. I was focusing a too narrow use case.

Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:245]: so ok, also considering we have already `SortedTopology` let's keep this as is and let's address possible cleanups/ghw extensions later

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 7 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go]: nope but I have just added some new unit test to cover it.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 8 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go]: done

File: cmd/performance-profile-creator/cmd/root.go (hunk 9 above)
Comment (jlojosnegros) [reply cmd/performance-profile-creator/cmd/root.go]: ok

File: test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go [hunk 12]
```diff
@@ -168,6 +172,43 @@ var _ = Describe("[rfe_id:OCP-38968][ppc] Performance Profile Creator", func() {
 		})
 	})
 
+	It("[test_id:XXXX1] Verify PPC fails when offlined cpu count is more than available cpus", func() {
```
Comment (jlojosnegros) [inline test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:175]: There is no "test_id" for these tests.
They are really simple tests, but I have left this here just to raise the question: Should I ask for a test_id or just ignore the tag?

Comment (fromanirh) [reply test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:175]: let's start ignoring the tag, we can add it later

File: test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go [hunk 13]
```diff
@@ -168,6 +172,43 @@ var _ = Describe("[rfe_id:OCP-38968][ppc] Performance Profile Creator", func() {
...
+		_, errData, _ := testutils.ExecAndLogCommandWithStderr(ppcPath, cmdArgs...)
+		ppcErrorString := errorStringParser(errData)
+		Expect(ppcErrorString).To(ContainSubstring("please specify the offlined CPU count in the range"))
+	})
+
+	It("[test_id:XXXX3] Verify PPC fails when the sum of offlined cpu count and reserved cpu count is more than available cpus", func() {
+		ppcArgs := []string{
+			fmt.Sprintf("--offlined-cpu-count=%d", 50),
+			fmt.Sprintf("--reserved-cpu-count=%d", 50),
+			"--rt-kernel=false",
```
Comment (fromanirh) [inline test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:203]: do we want e2e tests for:

- offline and reserved sets overlapping
- all the above with rt-kernel=true?

File: pkg/performanceprofile/profilecreator/profilecreator_test.go [hunk 14]
```diff
@@ -786,19 +786,50 @@ var _ = Describe("PerformanceProfileCreator: Test Helper cpuAccumulator", func()
...
 			cores := acc.Result().ToSlice()
 			Expect(cores).Should(Equal([]int{0, 1, 2}))
 		})
 
+		It("should not count elements already in the accumulator", func() {
+			acc := newCPUAccumulator()
+			node := topology1.Nodes[0]
+			n1, err := acc.AddCores(allCores, node.Cores)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(n1).To(BeNumerically("==", 4))
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator_test.go:1120]: or `To(Equal(4))` ?

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 15]
```diff
@@ -289,30 +304,147 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
+
+func (ghwHandler GHWHandler) GetOfflinedCPUs(alreadyUsedCPUs cpuset.CPUSet, cpuInfo *cpu.Info, topologyInfo *topology.Info, offlinedCPUCount int, disableHTFlag bool, htEnabled bool) (cpuset.CPUSet, error) {
+	offlinedCPUs := newCPUAccumulator()
+	lpOfflined := 0
+
+	//currently HT is enabled on the system and the user wants to disable HT
+	if htEnabled && disableHTFlag {
+		for _, node := range topologyInfo.Nodes {
+			acc, err := offlinedCPUs.AddCoresWithFilter(offlinedCPUCount, node.Cores, func(index, lpID int) bool {
+				return !alreadyUsedCPUs.Contains(lpID) && filterFirstLogicalProcessorInCore(index, lpID)
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:406]: do we have a e2e test which checks the CPUs picked by this logic and ensures we offlined only thread sibilings?

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 16]
```diff
@@ -289,30 +304,147 @@ func topologyHTDisabled(info *topology.Info) *topology.Info {
...
+				if err != nil {
+					return cpuset.CPUSet{}, err
+				}
+				lpOfflined += acc
+			}
+		}
+	} else {
+		for _, processor := range cpuInfo.Processors {
+			if processor.NumThreads <= uint32(offlinedCPUCount-lpOfflined) {
+				// We can offline a complete CPU socket
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator.go:427]: same, do we have a e2e test which ensures that we picked all the cpus pertaining a socket? e.g. nothing more? (and probably nothing less?)

File: pkg/performanceprofile/profilecreator/profilecreator_test.go [hunk 17]
```diff
@@ -404,6 +404,287 @@ var _ = Describe("PerformanceProfileCreator: Populating Reserved and Isolated CP
...
+			mustGatherDirAbsolutePath, err = filepath.Abs(mustGatherDirPath)
+			Expect(err).ToNot(HaveOccurred())
+			handle, err = NewGHWHandler(mustGatherDirAbsolutePath, node)
+			Expect(err).ToNot(HaveOccurred())
+			reservedCPUSet, isolatedCPUSet, offlinedCPUSet, err := handle.CalculateCPUSets(reservedCPUCount, offlinedCPUCount, splitReservedCPUsAcrossNUMA, disableHT)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(offlinedCPUSet.Intersection(reservedCPUSet).IsEmpty()).To(BeTrue())
+			Expect(offlinedCPUSet.Intersection(isolatedCPUSet).IsEmpty()).To(BeTrue())
+			Expect(offlinedCPUSet.Size()).To(BeNumerically("==", offlinedCPUCount))
```
Comment (fromanirh) [inline pkg/performanceprofile/profilecreator/profilecreator_test.go:460]: we should probalby add a check that the union of the returned set is the total set of the CPUs in the system

**Review by fromanirh — COMMENTED** [review]
/lgtm
/hold
Nice work! not blocking comments, let's see what other reviewers think

**Comment by fromanirh** [comment]
/approve
we surely want this change

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 16 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:427]: Yes and no .... 
There is a test addressing this case but to disable a complete socket is almost impossible right now. 
As reservedCPUs are calculated first and we do not have socket layout into account to do it, it is almost certain that at least one cpu of each socket would be reserved.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 15 above)
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:406]: ok, but we don't actually check in the system that cpuID=X is or is not sibling of cpuID=Y, right? meaning, we hardcoded the correct setting in the test, computed manually. Is not verified querying some data, is it?
Doing like this is OK, we can extend later.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 16 above)
Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:427]: anyway, we can hunt and use more system topologies if this helps us

**Review by marioferh — APPROVED** [review]
Tested and from the feature point of view seems right. We can discuss about the criteria to choose offlined cpus but for this PR is more than enough. Thanks @jlojosnegros

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 15 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:406]: Nope, we do not query any data in the tests, just compute the correct result manually.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 16 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:427]: > But we can reserve only cpus from socket 0 and offline all socket 1 in theory, can't we?

Surely we could do it, but right now we are not doing it at all. We do not consult CPU info to choose reserved CPUs, just TopologyInfo so we just choose the cpus in the same NUMA zone or balance them between the NUMA zones but we do not really know if they belong to the same or different sockets.

Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator.go:427]: ok, so we need to first clarify/unify our own terminology and then iterate from there. Even with multi-NUMA single-socket CPUs offlining an entire NUMA zone can yield significant improvements.

File: pkg/performanceprofile/profilecreator/profilecreator_test.go (hunk 17 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator_test.go:460]: I have been trying to add this but reserved + offlined + isolated = total is not always true. 

When `disable-ht` is `true` and reserved are calculated topology is modified deleting all the "sibling" logical processors so they could not be eligible when isolated is calculated. 

This modified topology is used to calculate isolated but not for the offflined because in fact in this case we really want to offline the sibling logical processors.

TL;DR
So, if `disable-ht` is `true` then `reserved + offlined + isolated = total` will be `false` **unless** offlined-cpu-count >= number of logical processor siblings

Comment (fromanirh) [reply pkg/performanceprofile/profilecreator/profilecreator_test.go:460]: That's a great testcase to add to document this behaviour :) This kind of knowledge should be explicitely recorded, better in tests (+ explanation comments where helpful)

Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator_test.go]: done

**Comment by fromanirh** [comment]
/hold cancel
lgtm is gone, no point in holding anymore

File: test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go (hunk 12 above)
Comment (jlojosnegros) [reply test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go]: done

File: test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go (hunk 13 above)
Comment (jlojosnegros) [reply test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:203]: I can add them but `rt-kernel` is not checked to calculate any of the cpuSets.

Comment (fromanirh) [reply test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:203]: ok, then no point. We're good as we are now.

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 18]
```diff
@@ -61,6 +61,21 @@ const (
 	allCores = -1
 )
 
+var (
+	noFilerLogicalProcessors = func(int, int) bool { return true }
```
Comment (yanirq) [inline pkg/performanceprofile/profilecreator/profilecreator.go:65]: s/noFilerLogicalProcessors/noFil**t**erLogicalProcessors

File: pkg/performanceprofile/profilecreator/profilecreator.go [hunk 19]
```diff
@@ -61,6 +61,21 @@ const (
...
+
+	// This filter is used to avoid offlining the first logical processor of each core.
+	// LogicalProcessors is a slice of ints representing the logical processor IDs assigned to
```
Comment (yanirq) [inline pkg/performanceprofile/profilecreator/profilecreator.go:68]: s/ints/integers

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 18 above)
Comment (yanirq) [reply pkg/performanceprofile/profilecreator/profilecreator.go:65]: other suggestion : change var name to unfilteredLogicalProcessors / allLogicalProcessors / LogicalProcessors

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 10 above)
Comment (yanirq) [reply pkg/performanceprofile/profilecreator/profilecreator.go:342]: this is an old behavior we already had.  we should indeed scan and sweep these messages into a different layer.

File: test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.json [hunk 20]
```diff
@@ -0,0 +1,8 @@
+{
+  "must-gather-dir-path": "must-gather.bare-metal",
+  "mcp-name": "worker-cnf",
+  "reserved-cpu-count": 4,
+  "offlined-cpu-count": 4,
+  "rt-kernel": true,
+  "power-consumption-mode": "low-latency"
+}
```
Comment (yanirq) [inline test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.json:8]: missing new line

File: test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1b.json [hunk 21]
```diff
@@ -0,0 +1,8 @@
+{
+  "must-gather-dir-path": "must-gather.bare-metal",
+  "mcp-name": "worker-cnf",
+  "reserved-cpu-count": 4,
+  "offlined-cpu-count": 40,
+  "rt-kernel": true,
+  "power-consumption-mode": "low-latency"
+}
```
Comment (yanirq) [inline test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1b.json:8]: missing new line

File: test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.json (hunk 20 above)
Comment (yanirq) [reply test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.json:8]: this is repetitive in all the json files added here.

File: test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.yaml [hunk 22]
```diff
@@ -0,0 +1,20 @@
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: performance
+spec:
+  cpu:
+    isolated: 1,3-39,41,43,45,47,49-51,53-79
+    reserved: 0,2,40,42
+    offlined: 44,46,48,52
```
Comment (yanirq) [inline test/e2e/performanceprofile/testdata/ppc-expected-profiles/profile1a.yaml:9]: @jlojosnegros   in all the expected profile nder this test data , do we have the `offlined` always set with values ? 
If the answer is yes then I would add at least one test profile without that field (since its optional) to test the default behavior.

File: test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go (hunk 13 above)
Comment (jlojosnegros) [reply test/e2e/performanceprofile/functests-performance-profile-creator/1_performance-profile_creator/ppc.go:203]: ok. the parameter is irrelevant but it is mandatory.

File: pkg/performanceprofile/profilecreator/profilecreator.go (hunk 10 above)
Comment (jlojosnegros) [reply pkg/performanceprofile/profilecreator/profilecreator.go:393]: Maybe would be better to scan and sweep in a different PR, is that ok ?

**Comment by jlojosnegros** [comment]
Minor changes and rebase over last master to avoid some CI problems.

**Comment by jlojosnegros** [comment]
rebasing over last master
